accord-core/src/main/java/accord/utils/async/AsyncResult.java

```diff
@@ -0,0 +1,96 @@
+package accord.utils.async;
+
+import java.util.concurrent.Executor;
+import java.util.function.BiConsumer;
+
+/**
+ * Handle for async computations that supports multiple listeners and registering
+ * listeners after the computation has started
+ */
+public interface AsyncResult<V>
+{
+    void listen(BiConsumer<? super V, Throwable> callback);
+
+    default void listen(BiConsumer<? super V, Throwable> callback, Executor executor)
+    {
+        listen(AsyncCallbacks.inExecutor(callback, executor));
+    }
+
+    default void listen(Runnable runnable)
+    {
+        listen((unused, failure) -> {
+            if (failure == null) runnable.run();
+            else throw new RuntimeException(failure);
+        });
+    }
+
+    default void listen(Runnable runnable, Executor executor)
+    {
+        listen(AsyncCallbacks.inExecutor(runnable, executor));
+    }
```

Comment:
consistency comment: can we rename to `addCallback`?  `AsyncChain` and C* `Future` use `addCallback`, so would be good be match